lib/job_queue/slurm_driver.cpp

```diff
@@ -118,22 +166,69 @@ bool slurm_driver_set_option( void * __driver, const char * option_key, const vo
     return true;
   }
 
+  if (strcmp(option_key, SLURM_PARTITION_OPTION) == 0) {
+    driver->partition = static_cast<const char*>(value);
+    return true;
+  }
+
   return false;
 }
 
 
 void slurm_driver_init_option_list(stringlist_type * option_list) {
+  stringlist_append_copy(option_list, SLURM_PARTITION_OPTION);
   stringlist_append_copy(option_list, SLURM_SBATCH_OPTION);
   stringlist_append_copy(option_list, SLURM_SCONTROL_OPTION);
   stringlist_append_copy(option_list, SLURM_SQUEUE_OPTION);
   stringlist_append_copy(option_list, SLURM_SCANCEL_OPTION);
 }
 
 
+static std::string make_submit_script(const char * cmd, int num_cpu, int argc, const char ** argv) {
+  char * submit        = (char*) util_alloc_tmp_file("/tmp" , "slurm-submit" , true);
+
+  FILE * submit_stream = util_fopen(submit, "w");
+  fprintf(submit_stream, "#!/bin/sh\n");
+
+  fprintf(submit_stream, "%s", cmd);  // Without srun?
+  for (int iarg=0; iarg < argc; iarg++)
+    fprintf(submit_stream, " %s", argv[iarg]);
+  fprintf(submit_stream, "\n");
+
+  fclose(submit_stream);
+  chmod(submit, S_IRWXU + S_IRGRP + S_IROTH);
+
+  std::string submit_script = submit;
+  free( submit );
+  return submit_script;
+}
+
+
+/*
+ The slurm jobs are submitted by first creating a submit script, which is a
+ small shell which contains the command to run along with possible slurm
+ options, and then this script is submitted with the 'sbatch' command.
+*/
+
 void * slurm_driver_submit_job( void * __driver, const char * cmd, int num_cpu, const char * run_path, const char * job_name, int argc, const char ** argv) {
   slurm_driver_type * driver = slurm_driver_safe_cast( __driver );
-  SlurmJob * job = nullptr;
-  return job;
+
+  auto submit_script = make_submit_script( cmd, num_cpu, argc, argv);
+  std::vector<std::string> sbatch_argv = {"--workdir=" + std::string(run_path), "--job-name=" + std::string(job_name), "--parsable"};
+  if (!driver->partition.empty())
+    sbatch_argv.push_back( "--partition=" + driver->partition );
+  sbatch_argv.push_back( submit_script );
+
+
+  auto file_content = load_stdout(driver->sbatch_cmd.c_str(), sbatch_argv);
+  util_unlink_existing( submit_script.c_str() );
+
+  int job_id = 0;
```

Comment:
Is this safe? Is it guaranteed that the job id will never be 0?

Reply:
It is at least guaranteed it will not be negative; I'll update